testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java

```diff
@@ -4435,13 +4436,40 @@ public void testRowLevelDelete()
         }
     }
 
+    @Test
+    public void verifySupportsRowLevelUpdateDeclaration()
+    {
+        if (hasBehavior(SUPPORTS_ROW_LEVEL_UPDATE) || hasBehavior(SUPPORTS_UPDATE)) {
```

Comment:
`|| hasBehavior(SUPPORTS_UPDATE)` is not needed

Reply:
But this query will not fail : 
`assertQueryFails("UPDATE " + table.getName() + " SET nationkey = 100 WHERE regionkey = 2", MODIFYING_ROWS_MESSAGE);`

If we support simple update and doesn't support row level update.
Like PhoenixConnector

Reply:
Oh so opposite then is true you should remove `hasBehavior(SUPPORTS_ROW_LEVEL_UPDATE)` because if something `SUPPORTS_UPDATE` then it `SUPPORTS_ROW_LEVEL_UPDATE`. So if you there should be connector that `SUPPORTS_ROW_LEVEL_UPDATE`, but not `SUPPORTS_UPDATE`. Am I right?

Reply:
In row level update we don't support expression like - SET nationkey = nationkey + 1, or update all values of row.

Reply:
Ok... let's get back to square one.

What  SUPPORTS_UPDATE means and what SUPPORTS_ROW_LEVEL_UPDATE means?

Reply:
```
 @Test
    public void verifySupportsRowLevelUpdateDeclaration()
    {
        if (hasBehavior(SUPPORTS_ROW_LEVEL_UPDATE)) {
        return 
         }

          assertQueryFails("UPDATE " + table.getName() + " SET nationkey = 100 WHERE regionkey = 2", MODIFYING_ROWS_MESSAGE);
```

- In this case Phoenix connector will not fail assertion because it support simple update and NOT row level update.

```
 @Test
    public void verifySupportsRowLevelUpdateDeclaration()
    {
        if (hasBehavior(SUPPORTS_UPDATE)) {
         return 
       }

          assertQueryFails("UPDATE " + table.getName() + " SET nationkey = 100 WHERE regionkey = 2", MODIFYING_ROWS_MESSAGE);
```

In this case Postgres will not fail assertion, because it supports row level update and NOT simple update

Reply:
I am asking about the semantics of these two fields. How does not match to `SUPPORTS_DELETE` and `SUPPORTS_ROW_LEVEL_DELETE`?

Reply:
> I am asking about the semantics of these two fields. How does not match to `SUPPORTS_DELETE` and `SUPPORTS_ROW_LEVEL_DELETE`

I think it because DELETES provides exactly the same functionality not subsets.

Reply:
> I am asking about the semantics of these two fields. How does not match to `SUPPORTS_DELETE` and `SUPPORTS_ROW_LEVEL_DELETE`?

and we don't have connectors which support DELETE but not ROW_LEVEL_DELETE and vice versa

Reply:
> SUPPORTS_UPDATE means, that we support full type of updates like constant, expression, update all - and it's separate pass in code.
>
> SUPPORTS_ROW_LEVEL_UPDATE means that it support only limited subset of update functionality like constant updates and it's separate pass in code.

I think it is opposite and it should be matching the thing we have with DELETE.

SUPPORTS_DELETE_ROW_LEVEL_UPDATE means that we do actually query execution execution and we could remove row by row, whereas SUPPORTS_DELETE means that connector supports delete and it does not say to what extent. If you SUPPORTS_DELETE=true and SUPPORTS_DELETE_ROW_LEVEL_UPDATE=false, then it mans only delete with metadata api (delete pushdown) is supported.

We need to follow the same.

CC: @hashhar  @findepi to double check me

Reply:
> If you SUPPORTS_DELETE=true and SUPPORTS_DELETE_ROW_LEVEL_UPDATE=false, then it mans only delete with metadata api (delete pushdown) is supported.

Ok, what should be combination of flags in case, we DON'T support delete with metadata api (delete pushdown), but support row level delete? SUPPORTS_DELETE=false and SUPPORTS_DELETE_ROW_LEVEL_UPDATE=true ?

Reply:
> what should be combination of flags in case, we DON'T support delete with metadata api (delete pushdown), but support row level delete? 

There should be no such state. If you support row level delete (or update) then you can delete (or update) anything. Metadata DELETE (or UPDATE) is just an optimization in such case.

Reply:
refactored